gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java

```diff
@@ -132,29 +149,113 @@ public static void renamePath(FileSystem fs, Path oldName, Path newName) throws
+        .contains(srcFs.getUri().getScheme()) && !S3_SCHEMES.contains(dstFs.getUri().getScheme())) {
+      renamePath(srcFs, src, dst);
+    } else {
+      copyPath(srcFs, src, dstFs, dst, true, false, conf);
     }
   }
 
   /**
-   * A wrapper around {@link HadoopUtils#renamePath(FileSystem, Path, Path)} and
-   * {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} which will rename the path
-   * if the src and dst filesystems are the same; otherwise, the src file will be moved the the dst filesystem.
-   * An {@link IOException} if {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} returns false.
+   * Copies a src {@link Path} from a srcFs {@link FileSystem} to a dst {@link Path} on a dstFs {@link FileSystem}. If
+   * either the srcFs or dstFs are S3 {@link FileSystem}s (as dictated by {@link #S3_SCHEMES}) then data is directly
+   * copied from the src to the dst. Otherwise data is first copied to a tmp {@link Path}, which is then renamed to the
+   * dst.
+   *
+   * @param srcFs the source {@link FileSystem} where the src {@link Path} exists
+   * @param src the {@link Path} to copy from the source {@link FileSystem}
+   * @param dstFs the destination {@link FileSystem} where the dst {@link Path} should be created
+   * @param dst the {@link Path} to copy data to
+   * @param tmp the temporary {@link Path} to use when copying data
+   * @param deleteSource true if the source should be deleted after the copy is complete, false otherwise
+   * @param overwriteDst true if the destination should be overwritten, false otherwise
+   * @param overwriteTmp true if the temp file should be overwritten, false otherwise
    */
-  public static void movePath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst) throws IOException {
-    if (srcFs.getUri().equals(dstFs.getUri())) {
-      renamePath(srcFs, src, dst);
+  public static void copyPath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst, Path tmp, boolean deleteSource,
+      boolean overwriteDst, boolean overwriteTmp, Configuration conf)
+      throws IOException {
+
+    if (S3_SCHEMES.contains(srcFs.getUri().getScheme()) || S3_SCHEMES.contains(dstFs.getUri().getScheme())) {
+      copyPath(srcFs, src, dstFs, dst, deleteSource, overwriteDst, conf);
     } else {
-      if (!FileUtil.copy(srcFs, src, dstFs, dst, true, false, dstFs.getConf())) {
-        throw new IOException(String.format("Failed to move %s to %s", src, dst));
+      if (!overwriteDst && dstFs.exists(dst)) {
+        throw new IOException(String.format("Cannot copy %s to %s because the dst exists", src, dst));
+      }
+      copyPath(srcFs, src, dstFs, tmp, false, overwriteTmp, conf);
+      try {
+        if (overwriteDst && dstFs.exists(dst)) {
+          deletePath(dstFs, dst, true);
+        }
+        renamePath(dstFs, tmp, dst);
+      } finally {
+        deletePath(dstFs, tmp, false);
```

Comment:
why should `recursive=false` here? If `src` is a directory it will throw IOException